src/coreclr/src/ToolBox/superpmi/mcs/removedup.cpp

```diff
@@ -0,0 +1,169 @@
+                if (md5HashMap != nullptr)
+                {
+                    // go through and delete items
+                    for (int j = 0; j < (int)md5HashMap->GetCount(); j++)
+                    {
+                        char* p = md5HashMap->GetItem(j);
+                        delete[] p;
+                    }
+                    delete md5HashMap;
+                }
+            }
+            delete m_inFile;
+            m_inFile = nullptr;
+        }
+        if (m_inFileLegacy != nullptr)
+        {
+            for (int i = 0; i < (int)m_inFileLegacy->GetCount(); i++)
+            {
+                DenseLightWeightMap<MethodContext*>* md5HashMap = m_inFileLegacy->GetItem(i);
+                if (md5HashMap != nullptr)
+                {
+                    // go through and delete items
+                    for (int j = 0; j < (int)md5HashMap->GetCount(); j++)
+                    {
+                        MethodContext* p = md5HashMap->GetItem(j);
+                        delete p;
+                    }
+                    delete md5HashMap;
+                }
+            }
+            delete m_inFileLegacy;
+            m_inFileLegacy = nullptr;
+        }
+    }
+}
+
+bool RemoveDup::unique(MethodContext* mc)
+{
+    if (m_inFile == nullptr)
+        m_inFile = new LightWeightMap<int, DenseLightWeightMap<char*>*>();
+
+    CORINFO_METHOD_INFO newInfo;
+    unsigned            newFlags = 0;
+    mc->repCompileMethod(&newInfo, &newFlags);
+
+    // Assume that there are lots of duplicates, so don't allocate a new buffer for the MD5 hash data
+    // until we know we're going to add it to the map.
+    char md5Buff[MD5_HASH_BUFFER_SIZE];
+    mc->dumpMethodMD5HashToBuffer(md5Buff, MD5_HASH_BUFFER_SIZE, /* ignoreMethodName */ true, &newInfo, newFlags);
+
+    if (m_inFile->GetIndex(newInfo.ILCodeSize) == -1)
+        m_inFile->Add(newInfo.ILCodeSize, new DenseLightWeightMap<char*>());
+
+    DenseLightWeightMap<char*>* ourRank = m_inFile->Get(newInfo.ILCodeSize);
+
+    for (unsigned i = 0; i < ourRank->GetCount(); i++)
+    {
+        char* md5Buff2 = ourRank->Get(i);
+        if (strncmp(md5Buff, md5Buff2, MD5_HASH_BUFFER_SIZE) == 0)
+        {
+            return false;
+        }
+    }
+
+    char* newmd5Buff = new char[MD5_HASH_BUFFER_SIZE];
+    memcpy(newmd5Buff, md5Buff, MD5_HASH_BUFFER_SIZE);
+    ourRank->Append(newmd5Buff);
+    return true;
+}
+
+bool RemoveDup::uniqueLegacy(MethodContext* mc)
+{
+    if (m_inFileLegacy == nullptr)
+        m_inFileLegacy = new LightWeightMap<int, DenseLightWeightMap<MethodContext*>*>();
+
+    CORINFO_METHOD_INFO newInfo;
+    unsigned            newFlags = 0;
+    mc->repCompileMethod(&newInfo, &newFlags);
+
+    if (m_inFileLegacy->GetIndex(newInfo.ILCodeSize) == -1)
+        m_inFileLegacy->Add(newInfo.ILCodeSize, new DenseLightWeightMap<MethodContext*>());
+
+    DenseLightWeightMap<MethodContext*>* ourRank = m_inFileLegacy->Get(newInfo.ILCodeSize);
```

Comment:
Maybe avoid the lookup here when `GetIndex()` returns -1 and set ourRank to the newly-constructed `DenseLightWeightMap`?